### ISSUE TITLE
Weight Loader v2 — protocol and migration plan (PR1–PR6)

### ISSUE THREAD

**Issue by JD-ETH** [sgl-project/sglang#31051]
# Weight Loader v2 — protocol and migration plan

Continuation of #24703. Foundation/reference path merged in #28671.

Audit basis: commit `3ebb7c2d0`. This plan includes native text loaders that fit the established dense/direct, standard-MoE or MLA protocols without adding a new shared loader abstraction. GLM-family loaders are isolated in a dedicated PR. Models requiring another protocol extension are explicitly not intended as part of this migration.

All migrations remain opt-in behind `SGLANG_ENABLE_WEIGHT_LOADER_V2` (default off).

---

### Complete and explicit

Every checkpoint tensor entering v2 must be loaded, intentionally skipped by a named/tested rule, or rejected as unexpected. Missing targets, incomplete expert shards and pending fusion inputs must fail before execution. Broad exception fallback around parameter loaders is prohibited.

### Load first, derive second

1. Load and verify checkpoint tensors.
2. Run `post_load_weights` for derived buffers, absorbed MLA weights, merged views or kernel preparation.

Every loading path must run the same post-load contract exactly once.

### Parent migration includes affected downstream paths

Audit inheritance, `super()`, unbound parent calls, bound delegates, reward/classification/embedding heads, MTP/NextN/Eagle drafts, registry aliases and multimodal wrappers embedding a migrated text model.

- **Bound delegate:** automatically enters v2; require env-off/env-on routing smoke.
- **Unbound parent call:** unsafe; fix in the parent PR.
- **Inherit/super:** migrate or verify in the parent PR.
- **Manual wrapper:** remains independent; do not claim v2 coverage.

The regression fixed by #31988 is the minimum failure mode every parent PR must guard against.

### Required tests

- CPU mapping tests for every stacked/expert/remap rule;
- legacy-vs-v2 state-dict or logits equivalence;
- PP filtering and EP/expert-shard completeness where applicable;
- quantized checkpoint coverage for differing names/layouts;
- affected wrapper/delegate smoke in the same PR;
- no silent drop, duplicate post-load call or iterator double-consumption.

---

## Planned coverage

Denominator: **187** unique files under `python/sglang/srt/models/` defining at least one `load_weights`; `models/utils.py` is excluded.

- **PR1:** 2 files; cumulative **2/187 (1.1%)**.
- **PR2:** 69 files; cumulative **71/187 (38.0%)**.
- **PR3:** 8 files; cumulative **79/187 (42.2%)**.
- **PR4:** 4 files; cumulative **83/187 (44.4%)**.
- **PR5:** 3 files; cumulative **86/187 (46.0%)**.
- **PR6:** 6 files; cumulative **92/187 (49.2%)**.

Accounting: **92 planned + 95 not intended in this migration = 187**. Kimi K3 files are absent from the denominator commit.

---

## PR1 — merged reference path

- `llama.py`
- `qwen2.py`

### Downstream impact

- **PR2 debt:** `llama_classification.py`, `llama_reward.py`, `llama_embedding.py`, `llama_eagle.py`, `llama_eagle3.py`, `mistral.py`, `mistral_eagle.py`, `qwen2_classification.py`, `qwen2_rm.py`, `qwen2_eagle.py`.
- **Delegated multimodal smoke:** `llava.py`, `llavavid.py`, `voxtral.py`, `phi4mm.py`, `glmasr.py`, `points_v15_chat.py`, `nvila.py`, `nvila_lite.py`, `midashenglm.py`, `locate_anything.py`, `dots_ocr.py`, applicable `minicpmv.py` paths.
- **Manual multimodal, independent:** Qwen2 VL/audio and Qwen2 paths in InternVL/InternS1/MiniCPM wrappers.
- **No-loader aliases, smoke only:** `teleflm.py`, `llama4.py`, `ministral3.py`.

---

## PR2 — dense/direct and standard MoE

Only model-local stacked/expert mappings, remaps, named skips, PP filtering and existing parameter loaders are permitted. No new shared protocol extension belongs in this PR.

### Dense/direct — 30 files

- `gemma2.py`, `granite.py`, `internlm2.py`, `olmo.py`, `olmo2.py`, `qwen3.py`
- `apertus.py`, `arcee.py`, `baichuan.py`, `commandr.py`
- `exaone.py`, `exaone4.py`, `gemma.py`, `gpt_bigcode.py`, `gpt_j.py`
- `hrm_text.py`, `iquest_loopcoder.py`, `jet_nemotron.py`
- `mimo.py`, `mimo_mtp.py`, `minicpm.py`, `nemotron_nas.py`
- `orion.py`, `phi3_small.py`, `qwen.py`, `sdar.py`
- `solar.py`, `stablelm.py`, `starcoder2.py`, `xverse.py`

### Standard MoE — 23 files

- `mixtral.py`, `olmoe.py`, `phimoe.py`, `qwen2_moe.py`, `qwen3_moe.py`
- `afmoe.py`, `bailing_moe.py`, `cohere2_moe.py`, `dbrx.py`, `deepseek.py`
- `ernie4.py`, `ernie4_eagle.py`, `exaone_moe.py`, `exaone_moe_mtp.py`
- `hunyuan.py`, `hunyuan_v3.py`, `hunyuan_v3_nextn.py`
- `laguna.py`, `minimax_m2.py`, `minimax_m3.py`, `mixtral_quant.py`
- `sdar_moe.py`, `xverse_moe.py`

### Mandatory wrappers/delegates — 16 files

- `gemma2_reward.py`, `internlm2_reward.py`, `granitemoe.py`
- `qwen3_classification.py`, `qwen3_embedding.py`, `qwen3_moe_mtp.py`
- `llama_classification.py`, `llama_reward.py`, `llama_embedding.py`
- `llama_eagle.py`, `llama_eagle3.py`, `mistral.py`, `mistral_eagle.py`
- `qwen2_classification.py`, `qwen2_rm.py`, `qwen2_eagle.py`

`qwen3_rm.py` has no own `load_weights`; it requires smoke but does not affect coverage.

### Deliverables

- `StackedParamsDispatch` and `ExpertParamsDispatch`;
- model-local aliases/remaps and named skips;
- PP filtering, tied-head behavior and required expert-shard completeness;
- wrapper preprocessing preserved before delegation;
- mapping matrices, wrapper smoke and dense/MoE equivalence.

### Downstream impact

- **Unbound—must fix:** Gemma2/InternLM2 reward, GraniteMoe and PR1 classification/reward/Eagle paths.
- **Inherit/super:** Qwen3-MoE MTP, Ernie Eagle, Exaone MTP, Hunyuan NextN and MiMo MTP.
- **No-loader smoke:** `mellum.py`, `qwen3_rm.py` and PR1 aliases.
- **Multimodal smoke:** applicable Cohere, Ernie, Gemma, Jet, Qwen, MiniCPM, SDAR and Xverse downstream wrappers; full towers remain outside coverage.

---

### Scope — 8 files

- `deepseek_v2.py`
- `deepseek_nextn.py`
- `mistral_large_3.py`
- `kimi_k25_eagle3.py`
- `longcat_flash.py`
- `longcat_flash_nextn.py`
- `minicpm3.py`
- `sarvam_moe.py`
- helper: `deepseek_common/deepseek_weight_loader.py` (not in denominator)

### Deliverables

- conditional shared-expert remap;
- explicit multi-input MLA fusion with pending-input verification;
- kv/FP8 remaps, PP/NextN filtering and exactly-once post-load derivation;
- async/stream-safe loading;
- fusion on/off, NextN, truncated/PP and equivalence tests.

### Downstream impact

- **Inherit/super:** DeepSeek NextN, Mistral Large 3 and `mistral_large_3_eagle.py` smoke.
- **Delegated multimodal smoke:** `deepseek_vl2.py`, `dots_vlm.py`, `kimi_k25.py`; verify stripped names, iterator ownership and post-load once.
- **Manual multimodal, independent:** `kimi_vl.py`, `deepseek_ocr.py`.
- LongCat and NextN migrate together. Kimi K2.5/Eagle3 is not Kimi K3.

---

## PR4 — Qwen3.5 and Qwen3-Next language paths

No additional model is added to PR4.

- `qwen3_5.py`
- `qwen3_next.py`
- `qwen3_5_mtp.py`
- `qwen3_next_mtp.py`

### Deliverables

- packed GDN and fused-expert dispatch;
- standard/shared-expert MoE and hybrid mixer loading for these Qwen families;
- MTP parity and strict text/vision key separation;
- dense/MoE, GDN, fused-expert and MTP equivalence/smoke tests.

### Downstream impact

- MiniCPM-V 4.6 text delegation requires smoke.
- Qwen3.5 conditional-generation classes remain manual multimodal paths.
- `interns2preview.py` requires smoke if the conditional loader changes.

---

## PR5 — DeepSeek V4 only

- `deepseek_v4.py`
- `deepseek_v4_nextn.py`
- `deepseek_v4_dspark.py`

No additional model was confirmed to share V4 compressor/`wo_a` semantics.

### Downstream impact

- NextN must preserve filtering and exactly-once post-load.
- DSPARK must preserve its draft path while sharing PR5 mappings.
- No additional downstream multimodal entry class was identified at `3ebb7c2d0`.

---

## PR6 — GLM family

GLM is isolated because one file can expose standard GLM loading while another entry class inherits the PR3 DeepSeek/MLA path. This PR depends on PR2 and PR3.

### Scope — 6 files

- `chatglm.py`
- `glm4.py`
- `glm4_moe.py`
- `glm4_moe_nextn.py`
- `glm4_moe_lite.py`
- `glm4_moe_lite_nextn.py`

### Deliverables

- ChatGLM version/prefix normalization and checked fused-QKV layout;
- GLM4 dense mappings and PP/tied-head behavior;
- GLM4-MoE shared-expert remap, expert completeness and NextN filtering;
- GLM4-MoE-Lite MLA reuse from PR3;
- DSA entry-class inheritance smoke across PR3 and PR6;
- family-wide env-off/env-on equivalence.

### Downstream impact

- `glm4v.py`, `glm4v_moe.py`, `glm_ocr.py`, `glm_ocr_nextn.py`, `glm_image_vl.py`, `glmasr.py` and `midashenglm.py` require routing smoke where they embed, inherit or manually duplicate a GLM text path.
- Full vision/OCR/audio tower loading remains outside coverage.

---

### Native text loaders requiring another protocol or separate acceptance — 28 files

- `opt.py` — one-source tensor fanout/TP reconstruction requires a transform/fanout protocol.
- `grok.py` — output multiplier conversion requires checked tensor transforms.
- `llada2.py` — lm-head normalization requires a value-transform hook.
- `dflash.py` — speculative draft-only loader requires separate draft acceptance.
- `dspark.py` — one-shot backbone/Markov/confidence routing requires a stream router and mergeable results.
- `gemma3_causal.py` — embedding variant requires a declared auxiliary checkpoint-source protocol.
- `gpt2.py` — Conv1D transpose requires checked tensor transforms.
- `persimmon.py` — reshape/transpose requires checked tensor transforms.
- `phi.py` — packed checkpoint fanout requires transform/fanout dispatch.
- `torch_native_llama.py` — alternate nested module tree requires recursive prefix dispatch and completeness accounting.
- `bailing_moe_linear.py`, `bailing_moe_nextn.py` — linear-attention routing requires a generic hybrid mixer protocol.
- `falcon_h1.py` — Mamba/convolution loading requires a hybrid mixer protocol.
- `gemma3n_causal.py` — nonstandard per-layer projections require hybrid dispatch.
- `kimi_linear.py` — KDA packing plus MLA derivation requires hybrid dispatch composed with PR3.
- `lfm2.py`, `lfm2_moe.py` — short-convolution and batched experts require hybrid and batched-expert dispatch.
- `nemotron_h.py`, `nemotron_h_mtp.py` — Mamba state/convolution plus MTP requires hybrid dispatch.
- `gemma4_causal.py`, `gemma4_mtp.py` — batched expert tensors/scales require batched-expert dispatch.
- `granitemoehybrid.py` — hybrid routing and combined expert tensors require hybrid plus batched-expert dispatch.
- `step3p5.py`, `step3p5_mtp.py` — batched experts/shared slots require batched-expert dispatch.
- `zaya.py` — fused experts and persistent CCA/balancing buffers require fanout and buffer-target accounting.
- `gpt_oss.py` — MXFP4/Quark layouts and deferred scales require quant-layout dispatch.
- `mimo_v2.py`, `mimo_v2_nextn.py` — TP resharding, packed QKV, DeepEP experts and auxiliary routing require multiple new protocols.

### Explicit model exclusion

- **Kimi K3 only:** `kimi_k3.py`, `kimi_k3_vl.py` — implementation/checkpoint contract is absent from the audited `main` commit; zero denominator files.

### Full multimodal/OCR/vision — 53 files

Cause: each file owns or interleaves a full tower loader. Bound text delegates still receive parent-PR smoke.

- `cohere2_vision.py`, `deepseek_janus_pro.py`, `deepseek_ocr.py`, `deepseek_vl2.py`, `dots_ocr.py`, `dots_vlm.py`, `ernie45_vl.py`
- `gemma3_mm.py`, `gemma3n_mm.py`, `gemma4_mm.py`, `gemma4_unified.py`
- `glm4v.py`, `glm4v_moe.py`, `glm_image_vl.py`, `glm_ocr.py`, `glm_ocr_nextn.py`
- `inkling.py`, `interns1.py`, `interns1pro.py`, `internvl.py`, `jet_vlm.py`
- `kimi_k25.py`, `kimi_vl.py`, `lfm2_vl.py`, `lightonocr.py`, `llava.py`, `llavavid.py`, `locate_anything.py`, `midashenglm.py`
- `minicpmo.py`, `minicpmv.py`, `minimax_m3_vl.py`, `mllama.py`, `mllama4.py`, `moss_vl.py`, `nano_nemotron_vl.py`, `nvila.py`, `nvila_lite.py`
- `paddleocr_vl.py`, `phi4mm.py`, `pixtral.py`, `points_v15_chat.py`
- `qwen2_5_vl.py`, `qwen2_vl.py`, `qwen3_omni_moe.py`, `qwen3_vl.py`, `qwen3_vl_moe.py`
- `sarashina2_vision.py`, `step3_vl.py`, `step3_vl_10b.py`, `step3p7.py`, `unlimited_ocr.py`, `yivl.py`

### Speech/audio-specific — 7 files

Cause: audio encoder/decoder contracts and tests are required.

- `glmasr.py`, `mimo_v2_asr.py`, `parakeet.py`, `qwen2_audio.py`, `qwen3_asr.py`, `voxtral.py`, `whisper.py`

### Encoder-only — 5 files

Cause: encoder/classification loading requires a separate acceptance protocol.

- `bert.py`, `clip.py`, `radio.py`, `roberta.py`, `siglip2.py`

### Backend adapters — 2 files

Cause: framework adapters are not native model-family migrations.

- `mindspore.py`, `transformers.py`

---

## Dependencies and merge order

```text
PR1 merged
  └─ PR2 dense/direct + standard MoE + wrappers
       ├─ PR3 MLA + wrappers
       └─ PR4 Qwen3.5/Qwen3-Next + wrappers

PR5 DeepSeek V4
  └─ follows stabilization of PR3 post-load/MLA contracts

PR6 GLM family
  └─ depends on PR2 standard dispatch and PR3 MLA/DSA behavior
```

---

## Acceptance checklist

- [x] PR1: 2 reference files merged (#28671)
- [ ] Completeness result and verification contract
- [ ] PR2: 69 dense/direct, standard-MoE and wrapper files
- [ ] PR3: 8 MLA files
- [ ] PR4: 4 Qwen3.5/Qwen3-Next files; no additional models
- [ ] PR5: 3 DeepSeek V4 files
- [ ] PR6: 6 GLM-family files
- [ ] Every checkpoint tensor loaded, explicitly skipped or rejected
- [ ] Every expert shard and fusion input verified complete
- [ ] `post_load_weights` exactly once after successful verification
- [ ] Env-off parity and env-on equivalence for every parent/wrapper
- [ ] Models requiring new protocol extensions remain outside this migration
- [ ] Kimi K3 remains the only explicit model exclusion

---

## References

- #28671 — merged foundation/demo
- #31988 — wrapper regression caused by unbound parent dispatch
- #24703 — earlier auto-loader discussion

**Comment by JD-ETH** [comment]
## PR2–PR6 stack (draft)

Graphite `gt submit` blocked (repo not synced in Graphite for this fork). Branches pushed to `JD-ETH/sglang`; draft PRs opened:

| Slice | Fork PR (relative base) | Upstream draft (vs `main`) |
|-------|-------------------------|----------------------------|
| PR1 | — | https://github.com/sgl-project/sglang/pull/28671 |
| PR2 | https://github.com/JD-ETH/sglang/pull/11 | (see new PR) |
| PR3 | https://github.com/JD-ETH/sglang/pull/12 | (see new PR) |
| PR4 | https://github.com/JD-ETH/sglang/pull/13 | (see new PR) |
| PR5 | https://github.com/JD-ETH/sglang/pull/14 | (see new PR) |
| PR6 | https://github.com/JD-ETH/sglang/pull/15 | (see new PR) |

Local Graphite stack: `PR1 → PR2 → PR3 → PR4 → PR5 → PR6`. Unit tests on tip: 28 passed.

**Comment by JD-ETH** [comment]
### Upstream draft PRs (vs `main`)

| Slice | Upstream | Fork (stacked base) |
|-------|----------|---------------------|
| PR1 | https://github.com/sgl-project/sglang/pull/28671 | — |
| PR2 | https://github.com/sgl-project/sglang/pull/31264 | https://github.com/JD-ETH/sglang/pull/11 |
| PR3 | https://github.com/sgl-project/sglang/pull/31265 | https://github.com/JD-ETH/sglang/pull/12 |
| PR4 | https://github.com/sgl-project/sglang/pull/31266 | https://github.com/JD-ETH/sglang/pull/13 |
| PR5 | https://github.com/sgl-project/sglang/pull/31267 | https://github.com/JD-ETH/sglang/pull/14 |
| PR6 | https://github.com/sgl-project/sglang/pull/31268 | https://github.com/JD-ETH/sglang/pull/15 |

Use **fork PRs** for relative per-slice diffs. Upstream drafts are cumulative vs `main` until parents merge. `gt submit` needs Graphite repo sync for this fork.
